pkg/controller/backup_sync_controller.go

```diff
@@ -21,378 +21,408 @@ import (
 	backupStoreGetter persistence.ObjectBackupStoreGetter,
-	logger logrus.FieldLogger,
-) Interface {
-	if syncPeriod <= 0 {
-		syncPeriod = time.Minute
-	}
-	logger.Infof("Backup sync period is %v", syncPeriod)
-
-	c := &backupSyncController{
-		genericController:       newGenericController(BackupSync, logger),
-		backupClient:            backupClient,
-		kbClient:                kbClient,
-		podVolumeBackupClient:   podVolumeBackupClient,
+	logger logrus.FieldLogger) *backupSyncReconciler {
+	return &backupSyncReconciler{
+		client:                  client,
 		namespace:               namespace,
-		defaultBackupLocation:   defaultBackupLocation,
-		defaultBackupSyncPeriod: syncPeriod,
-		backupLister:            backupLister,
-		csiVSLister:             csiVSLister,
-		csiSnapshotClient:       csiSnapshotClient,
-		kubeClient:              kubeClient,
-
-		// use variables to refer to these functions so they can be
-		// replaced with fakes for testing.
-		newPluginManager:  newPluginManager,
-		backupStoreGetter: backupStoreGetter,
+		defaultBackupSyncPeriod: defaultBackupSyncPeriod,
+		newPluginManager:        newPluginManager,
+		backupStoreGetter:       backupStoreGetter,
+		logger:                  logger,
 	}
+}
 
-	c.resyncFunc = c.run
-	c.resyncPeriod = 30 * time.Second
+// Reconcile syncs between the backups in cluster and backups metadata in object store.
+func (b *backupSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := b.logger.WithField("controller", BackupSync)
+	log.Debug("Checking for existing backup storage locations to sync into cluster.")
 
-	return c
-}
+	location := &velerov1api.BackupStorageLocation{}
+	err := b.client.Get(ctx, req.NamespacedName, location)
+	if err != nil {
+		log.WithError(err).Errorf("BackupStorageLocation %s is not found", req.NamespacedName.String())
+		return ctrl.Result{Requeue: false}, err
+	}
 
-// orderedBackupLocations returns a new slice with the default backup location first (if it exists),
-// followed by the rest of the locations in no particular order.
-func orderedBackupLocations(locationList *velerov1api.BackupStorageLocationList, defaultLocationName string) []velerov1api.BackupStorageLocation {
-	var result []velerov1api.BackupStorageLocation
+	pluginManager := b.newPluginManager(log)
+	defer pluginManager.CleanupClients()
 
-	for i := range locationList.Items {
-		if locationList.Items[i].Name == defaultLocationName {
-			// put the default location first
-			result = append(result, locationList.Items[i])
-			// append everything before the default
-			result = append(result, locationList.Items[:i]...)
-			// append everything after the default
-			result = append(result, locationList.Items[i+1:]...)
+	log = log.WithField("backupLocation", location.Name)
+
+	syncPeriod := b.defaultBackupSyncPeriod
+	if location.Spec.BackupSyncPeriod != nil {
+		syncPeriod = location.Spec.BackupSyncPeriod.Duration
+		if syncPeriod == 0 {
```

Comment:
How about putting the checking logic into a FilterFunc of `PeriodicalEnqueueSource` so that no reconciling will be triggered?

Reply:
Sure.